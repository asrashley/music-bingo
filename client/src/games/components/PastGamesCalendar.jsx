import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { reverse } from 'named-urls';

import { PastGamesThemePropType } from '../types/PastGamesThemeFields';
import routes from '../../routes';

function CalendarTableHeader({ months }) {
    const splitMonths = months.map(key => key.split('-'));
    const years = [];
    let lastYear = null;
    for (const [year] of splitMonths) {
        if (lastYear !== year) {
            years.push({ year, colSpan: 1 });
            lastYear = year;
        } else {
            years[years.length - 1].colSpan += 1;
        }
    }

    return <thead>
        <tr className="year-row">
            <th rowSpan={2} className="theme-title">Theme</th>
            {years.map(({ year, colSpan }) => <th key={year} colSpan={colSpan} className="year">{year}</th>)}
            <th rowSpan={2} className="theme-title">Theme</th>
        </tr>
        <tr className="month-row">
            {splitMonths.map(([year, month]) => <th key={`m-${year}-${month}`} className="month">{month}</th>)}
        </tr>
    </thead>
}
CalendarTableHeader.propTypes = {
    months: PropTypes.arrayOf(PropTypes.string),
}

function CalendarCell({ value }) {
    const val = value ?? 0;
    const className = val ? 'calendar-cell active' : 'calendar-cell inactive';
    return <td className={className}>{val}</td>
}
CalendarCell.propTypes = {
    value: PropTypes.number,
};

function TitleCell({ slug, title }) {
    return (<td className="theme-title">
        <Link to={reverse(`${routes.pastGamesByTheme}`, { slug })}>
            {title}
        </Link>
    </td>);
}
TitleCell.propTypes = {
    slug: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
}

function ThemeRow({ months, theme }) {
    const { slug, row, title } = theme;
    return (<tr>
        <TitleCell key={`before-${slug}`} slug={slug} title={title} />
        {months.map(name => <CalendarCell key={name} value={row[name]} />)}
        <TitleCell key={`after-${slug}`} slug={slug} title={title} />
    </tr>)
}
ThemeRow.propTypes = {
    theme: PastGamesThemePropType.isRequired,
    months: PropTypes.arrayOf(PropTypes.string),
};

export function PastGamesCalendar({ themes, months, monthsMap }) {
    return (<table className="past-games-calendar table table-striped">
        <CalendarTableHeader months={months} />
        <tbody>
            {themes.map((theme) => <ThemeRow theme={theme} key={theme.slug} months={months} />)}
        </tbody>
    </table>);
}
PastGamesCalendar.propTypes = {
    themes: PropTypes.arrayOf(PastGamesThemePropType),
    months: PropTypes.arrayOf(PropTypes.string),
};

export const PastGamesCalendarPropType = PropTypes.shape({
    themes: PropTypes.arrayOf(PastGamesThemePropType),
    months: PropTypes.arrayOf(PropTypes.string),
});