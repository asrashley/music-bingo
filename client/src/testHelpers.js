import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import log from 'loglevel';
import { screen, fireEvent } from '@testing-library/react';
import waitForExpect from 'wait-for-expect';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import { history } from './store/history';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ConnectedRouter history={history}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ConnectedRouter>
    </Provider>);
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
};

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    if (modFn !== undefined) {
      return jsonResponse(modFn(url, data['default']));
    }
    return jsonResponse(data['default']);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = (url, opts) => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async (url, opts) => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    return jsonResponse(data['default']);
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    const userIsValid = userDatabase.some(item => (
      item.username === username && item.password === password));
    log.debug(`loginUser: username="${username}" password="${password}" valid=${userIsValid}`);
    if (!userIsValid) {
      return 401;
    }
    const data = await import('./fixtures/user.json');
    data['default'].refreshToken = refreshToken;
    data['default'].accessToken = getAccessToken();
    loggedIn = true;
    return jsonResponse(data['default']);
  };
  const logoutUser = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  fetchMock.config.fallbackToNetwork = false;
  fetchMock.config.warnOnFallback = true;
  log.debug(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields) {
  fields.forEach(({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    fireEvent.input(screen.getByLabelText(label, { exact }), {
      target: {
        value
      }
    });
  });
  await waitForExpect(() => {
    const last = fields[fields.length - 1];
    const { value } = document.querySelector(`input[name="${last.label}"`);
    expect(value).toBe(fields[fields.length - 1].value);
  });
}

