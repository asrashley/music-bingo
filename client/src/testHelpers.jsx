import { render, queries, buildQueries, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import { ReduxRouter } from '@lagunovsky/redux-react-router';
import log from 'loglevel';
import { Readable } from "stream";
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';

import { DisplayDialog } from './components/DisplayDialog';
import { createStore } from './store/createStore';
import fetchMockHandlerTest from 'fetch-mock/esm/client';
import jestify from 'fetch-mock-jest/jestify';

import userData from './fixtures/user.json';

export const fetchMock = jestify(fetchMockHandlerTest);

/* custom query that looks for a "data-last-update" attribute in the container and compares that to
  the value provided to the query. It can be used to check if a component has been updated or
  to wait for it to re-render
  */
const queryAllByLastUpdate = (container, lastUpdate, options = {}) => {
  const { comparison } = {
    comparison: 'equals',
    ...options
  };
  log.trace(`queryAllByLastUpdate lastUpdate=${lastUpdate} comparison=${comparison} container="${container.nodeName}.${container.className}" ${container.id}`);
  return Array.from(container.querySelectorAll('[data-last-update]'))
    .filter((elt) => {
      const update = parseInt(elt.dataset.lastUpdate, 10);
      log.trace(`${elt.nodeName}.${elt.className} lastUpdate = ${update}`);
      if (isNaN(update)) {
        return false;
      }
      switch (comparison) {
        case 'equals':
          return update === lastUpdate;
        case 'greaterThan':
          return update > lastUpdate;
        case 'greaterThanOrEquals':
          return update >= lastUpdate;
        case 'lessThan':
          return update < lastUpdate;
        case 'lessThanOrEquals':
          return update <= lastUpdate;
        default:
          return true;
      }
    });
};
export const [queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate] = buildQueries(
  queryAllByLastUpdate,
  (container, selector) => `Found multiple elements from ${container} with last update selector: ${selector}`,
  (container, selector, opts = {}) => {
    const { comparison = 'equals' } = opts;
    return (`Unable to find an element from ${container.nodeName}.${container.className} with last update ${comparison} ${selector}`);
  }
);
const lastUpdatedQueries = { queryByLastUpdate, getAllLastUpdate, getByLastUpdate, findAllLastUpdate, findByLastUpdate };

/* custom query that uses a CSS selector to find elements */
const queryAllBySelector = (container, selector) => {
  log.trace(`queryAllBySelector ${selector}`);
  return Array.from(container.querySelectorAll(selector));
};
const [queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector] = buildQueries(
  queryAllBySelector,
  (container, selector) => `Found multiple elements from ${container} with selector: ${selector}`,
  (container, selector) => `Unable to find an element from ${container} with selector: ${selector}`,
);
const bySelectorQueries = { queryBySelector, getAllBySelector, getBySelector, findAllBySelector, findBySelector };

const routerSelector = (state) => state.router

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    history = createMemoryHistory(),
    store = createStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (<Provider store={store}>
      <ReduxRouter history={history} routerSelector={routerSelector}>
        <DisplayDialog>
          {children}
        </DisplayDialog>
      </ReduxRouter>
    </Provider>);
  }
  Wrapper.propTypes = {
    children: PropTypes.node,
  };
  const events = userEvent.setup({
    delay: 10,
  });
  return {
    events,
    store,
    ...render(ui, {
      wrapper: Wrapper,
      queries: {
        ...queries,
        ...lastUpdatedQueries,
        ...bySelectorQueries
      },
      ...renderOptions
    })
  };
}

const protectedRoutes = {
  '/api/directory': true,
  '/api/game': true,
  '/api/settings': false,
  '/api/user': true,
};

export function jsonResponse(payload, status = 200) {
  const body = JSON.stringify(payload);
  return {
    body,
    status,
    headers: {
      'Cache-Control': 'max-age = 0, no_cache, no_store, must_revalidate',
      'Content-Type': 'application/json',
      'Content-Length': body.length
    }
  };
}

export class MockResponse extends Response {
  constructor(...args) {
    if (args[0] instanceof ReadableStream) {
      args[0] = Readable.from(args[0]);
    }
    super(...args);
  }
}

export function installFetchMocks(fetchMock, {
  loggedIn = false,
  currentAccessToken = 1,
  refreshToken = "refresh.token"
} = {}) {
  const responseModifiers = {};
  let serverStatus = null;
  const apiRequest = async (url, opts) => {
    log.trace(`apiRequest ${url}`);
    if (serverStatus !== null) {
      return serverStatus;
    }
    if (protectedRoutes[url]) {
      const { headers } = opts;
      log.trace(`Authorization = headers?.Authorization`);
      const bearer = `Bearer ${getAccessToken()}`;
      if (headers?.Authorization !== bearer) {
        log.debug(`Invalid access token "${headers?.Authorization}" !== "${bearer}"`);
        return 401;
      }
    }
    const filename = url.replace(/^\/api/, './fixtures');
    log.trace(`load ${filename}`);
    const data = await import(`${filename}.json`);
    log.trace(`${filename} = ${Object.keys(data['default']).join(',')}`);
    if (url === '/api/settings' && !loggedIn) {
      data['default'] = {
        privacy: data['default'].privacy
      };
    }
    const modFn = responseModifiers[url];
    let responsePayload = data['default'];
    if (modFn !== undefined) {
      log.debug(`modify response for ${url}`);
      responsePayload = modFn(url, responsePayload);
      log.debug(`modify response for ${url} done`);
    }
    return jsonResponse(responsePayload);
  };
  const getAccessToken = () => `access.token.${currentAccessToken}`;
  const refreshAccessToken = () => {
    currentAccessToken++;
    log.trace(`refreshAccessToken ${currentAccessToken}`);
    return {
      'accessToken': getAccessToken()
    };
  };
  const checkUser = async () => {
    log.debug(`checkUser loggedIn=${loggedIn}`);
    if (serverStatus !== null) {
      log.debug(`checkUser serverStatus=${serverStatus}`);
      return serverStatus;
    }
    if (!loggedIn) {
      return 401;
    }
    return jsonResponse({
      ...userData,
      refreshToken,
      accessToken: getAccessToken(),
    });
  };
  const userDatabase = [{
    username: 'user',
    email: 'a.user@example.tld',
    password: 'mysecret'
  }];
  const loginUser = async (url, opts) => {
    if (serverStatus !== null) {
      log.debug(`loginUser status=${serverStatus}`);
      return serverStatus;
    }
    const { username, password } = JSON.parse(opts.body);
    let index = -1;
    let dbEntry;
    userDatabase.forEach((item, i) => {
      if (item.username === username && item.password === password) {
        dbEntry = item;
        index = i;
      }
    });
    log.debug(`loginUser: username="${username}" password="${password}" index=${index}`);
    if (!dbEntry) {
      return 401;
    }
    const user = {
      ...userData,
      pk: index + 2,
      ...dbEntry,
      refreshToken,
      accessToken: getAccessToken()
    };
    delete user.password;
    loggedIn = true;
    return jsonResponse(user);
  };
  const logoutUser = () => {
    if (serverStatus !== null) {
      log.trace(`logoutUser status=${serverStatus}`);
      return serverStatus;
    }
    log.trace('logoutUser');
    loggedIn = false;
    return jsonResponse('Logged out');
  };
  const checkIfUserExists = (url, opts) => {
    if (serverStatus !== null) {
      log.trace(`checkIfUserExists status=${serverStatus}`);
      return serverStatus;
    }
    const { username, email } = JSON.parse(opts.body);
    const response = {
      "username": false,
      "email": false
    };
    userDatabase.forEach((item) => {
      if (item.username === username) {
        response.username = true;
      }
      if (item.email === email) {
        response.email = true;
      }
    });
    log.debug(`checkIfUserExists username=${username} email=${email} response=${JSON.stringify(response)}`);
    return jsonResponse(response);
  };

  Object.assign(fetchMock.config, {
    fallbackToNetwork: false,
    warnOnFallback: true,
    Response: MockResponse,
  });
  log.trace(`installFetchMocks ${loggedIn}`);
  fetchMock
    .get('/api/directory', apiRequest)
    .get(/api\/directory\/\d+/, apiRequest)
    .get('/api/games', apiRequest)
    .get(/api\/game\/\d+/, apiRequest)
    .post('/api/refresh', refreshAccessToken)
    .get('/api/settings', apiRequest)
    .get('/api/user', checkUser)
    .post('/api/user', loginUser)
    .delete('/api/user', logoutUser)
    .post('/api/user/check', checkIfUserExists)
    .get('/api/users', apiRequest)
    .get('/api/user/guest', apiRequest);

  return {
    getAccessToken,
    getRefreshToken: () => refreshToken,
    isLoggedIn: () => loggedIn,
    addUser: (user) => userDatabase.push(user),
    logout: () => loggedIn = false,
    setResponseModifier: (url, fn) => {
      responseModifiers[url] = fn;
    },
    setServerStatus: (code) => {
      serverStatus = code;
    },
    jsonResponse
  };
}

export async function setFormFields(fields, events) {
  await Promise.all(fields.map(async ({ label, value, exact }) => {
    if (exact === undefined) {
      exact = true;
    }
    const elt = await screen.findByLabelText(label, { exact });
    if (events) {
      //await events.click(elt);
      await waitFor(() => events.clear(elt));
      /*
            await waitFor(() => events.clear(elt));
        await waitFor(() => events.type(elt, value));
      }*/
    }
    await waitFor(() => fireEvent.input(elt, {
      target: {
        value
      }
    }));
  }));
  const { label, exact, value } = fields[fields.length - 1];
  const elt = await screen.findByLabelText(label, { exact });
  await waitFor(() => {
    expect(elt.value).toBe(value);
  });
}

