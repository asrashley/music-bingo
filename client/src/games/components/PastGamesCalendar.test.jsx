import { renderWithProviders } from '../../testHelpers';
import { PastGamesCalendar } from './PastGamesCalendar';

const themes = [{
    "key": "100nows",
    "title": "100 Nows",
    "row": {
        "2020-02": 1,
        "2020-08": 1,
        "2022-04": 1,
        "2022-11": 1,
        "2023-09": 1
    },
    "lastUsed": new Date("2023-09-21T17:52:25.041Z"),
    "elapsedTime": 5795103227
}, {
    "key": "1973",
    "title": "1973",
    "row": {
        "2023-10": 1
    },
    "lastUsed": new Date("2023-10-05T16:44:53.615Z"),
    "elapsedTime": 4589554653
}, {
    "key": "2000s",
    "title": "2000s",
    "row": {
        "2018-08": 1,
        "2019-06": 1,
        "2020-01": 1,
        "2021-11": 1,
        "2022-08": 1,
        "2023-05": 1
    },
    "lastUsed": new Date("2023-05-18T09:49:39.188Z"),
    "elapsedTime": 16710469080
}];

describe('PastGamesCalendar component', () => {
    beforeAll(() => {
        jest.useFakeTimers('modern');
        jest.setSystemTime(new Date('08 Feb 2023 10:12:00 GMT').getTime());
    });

    afterAll(() => jest.useRealTimers());

    it('matches snapshot', () => {
        const monthKeys = new Set();
        themes.forEach((theme) => {
            Object.keys(theme.row).forEach(cal => monthKeys.add(cal));
        });
        const months = [...monthKeys].sort();
        const { asFragment } = renderWithProviders(<PastGamesCalendar themes={themes} months={months} />);
        expect(asFragment()).toMatchSnapshot();
    });

});