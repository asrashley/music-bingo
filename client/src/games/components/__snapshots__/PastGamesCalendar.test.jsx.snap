// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PastGamesCalendar component matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="past-games-calendar table table-striped"
  >
    <thead>
      <tr
        class="year-row"
      >
        <th
          class="theme-title"
          rowspan="2"
        >
          Theme
        </th>
        <th
          class="year"
          colspan="1"
        >
          2018
        </th>
        <th
          class="year"
          colspan="1"
        >
          2019
        </th>
        <th
          class="year"
          colspan="3"
        >
          2020
        </th>
        <th
          class="year"
          colspan="1"
        >
          2021
        </th>
        <th
          class="year"
          colspan="3"
        >
          2022
        </th>
        <th
          class="year"
          colspan="3"
        >
          2023
        </th>
        <th
          class="theme-title"
          rowspan="2"
        >
          Theme
        </th>
      </tr>
      <tr
        class="month-row"
      >
        <th
          class="month"
        >
          08
        </th>
        <th
          class="month"
        >
          06
        </th>
        <th
          class="month"
        >
          01
        </th>
        <th
          class="month"
        >
          02
        </th>
        <th
          class="month"
        >
          08
        </th>
        <th
          class="month"
        >
          11
        </th>
        <th
          class="month"
        >
          04
        </th>
        <th
          class="month"
        >
          08
        </th>
        <th
          class="month"
        >
          11
        </th>
        <th
          class="month"
        >
          05
        </th>
        <th
          class="month"
        >
          09
        </th>
        <th
          class="month"
        >
          10
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/100nows"
          >
            100 Nows
          </a>
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/100nows"
          >
            100 Nows
          </a>
        </td>
      </tr>
      <tr>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/1973"
          >
            1973
          </a>
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/1973"
          >
            1973
          </a>
        </td>
      </tr>
      <tr>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/2000s"
          >
            2000s
          </a>
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell active"
        >
          1
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="calendar-cell inactive"
        >
          0
        </td>
        <td
          class="theme-title"
        >
          <a
            href="/history/themes/2000s"
          >
            2000s
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
